### PR TITLE
MM-30212 Support dynamic autocomplete for built-in slash commands

### PR DESCRIPTION
**Pull request by wiggin77**
#### Summary
This PR adds support for dynamic auto-completion lists for built-in slash commands.  

A moniker within the fetch url determines whether the request should be sent via HTTP(S) or routed internally to a built-in command.  The moniker is of the form `builtin:` followed by the name of the built-in command to use.  e.g. `builtin:share` would be become a call to the `share` command to fetch auto-completion items.  

Command providers implement the optional `DynamicListProvider` interface to indicate they can provide dynamic auto-completion.

This is required by the upcoming Shared Channels feature (https://docs.google.com/document/d/12Jwu9aODzVArNP3z6_ivTePIvz9EsjDpwSSVmzl5pU8/edit?usp=sharing).

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-30212

#### Release Note

```release-note
Dynamic auto-completion now supported for built-in slash commands. 
```

### CODE REVIEW

File: app/command_autocomplete.go [hunk 1]
```diff
@@ -12,6 +14,11 @@ import (
 	"github.com/mattermost/mattermost-server/v5/model"
 )
 
+// DynamicListProvider dynamically provides auto-completion for built-in commands.
+type DynamicListProvider interface {
```
Comment (mickmister) [inline app/command_autocomplete.go:18]: Can we incorporate a sense of `Autocomplete` here? Possibly `AutocompleteDynamicArgProvider`?

Comment (wiggin77) [reply app/command_autocomplete.go]: Sounds good.  Done via https://github.com/mattermost/mattermost-server/pull/16196/commits/ccb5167c30a853d6f9732858e780cb74a756b381

**Review by mickmister — APPROVED** [review]
Functionality LGTM. 0/5 on the naming

Edit: Naming looks good as well :+1:

**Review by iomodo — APPROVED** [review]
Looks great! Thanks @wiggin77 !

**Comment by wiggin77** [comment]
Hi @aaronrothschild. Any objections to this getting merged?  Also, who typically QA's slash commands?

**Review by aaronrothschild — APPROVED** [review]
LGTM

**Comment by wiggin77** [comment]
Hi @lindalumitchell, this is to add autocomplete to built-in slash commands.  So far the only command using it is an upcoming set of slash commands for shared channels.  I suggest deferring QA until it can be tested with those.  Also, there are comprehensive unit tests already.

**Review by lindalumitchell — APPROVED** [review]

**Pull request merged**